README.md
# RNN-MTLBlog-Title-Generator
An LSTM RNN that learns from a corpus of MTLBlog titles and generates similar titles

Run learner.py to construct a model first, then run generator.py to output titles.


news.txt
Trump Might Have Finally Declared Himself To Be Everyone's Favourite President
Montreal Doctors Mysteriously Disappears Leaving Behind 1500 Patients
What Would Happen If "The New Yorker" Were Set In Montreal
China's "Alien Hunting Telescope" Is Making Some Amazing Discoveries
Justin Trudeau Wants Canadian Men To Do This For Women
Montreal's La Ronde Announces Insane $50 Season Passes For Summer 2018
Thanks To The Internet, Millennials Are Ruining Christmas
Thousands Of People Were In Awe As Canada’s “Christmas Train” Arrived In Montreal (Photos)
The Quebec Government Is Officially Doing More For Anglophones
Quebec Wants To Build A Monorail From Montreal To Quebec City
Montreal New "Ugly Christmas Tree" Is Actually Amazing (Photos)
Canadians Are More In Debt Than Anyone Else In The World
Montreal Has Become A Mecca Of Video Games
Medical Cannabis Company To Open In West Island
This Website Tells You How Many People Died So Far On Black Friday
Controversy Surrounding Suspicious Death Of Immigrant In Montreal Police Custody
Quebec Government's Efforts To Teach French To Immigrants Has Failed
Air Canada Is Having A Massive 50% Off Black Friday Sale Right Now
Quebec Premier Criticizes Adidas Manager For Speaking In English
How A Canadian Court Case May Ruin Quebec's Plan For Legal Marijuana
Downtown Montreal Is Getting A $200 Million Makeover
Top 20 Most Dangerous Cities In Canada
Cuba Forms Alliance With North Korea Against Trump
Video Showing Truck On Fire On Montreal's Highway 40
Canada Begins Rolling Out Cancer Warnings On Alcohol
Montreal Ranked #1 Most Affordable City To Travel To In The World
A US-Bound North Korean Missile Could Fly Over Montreal
Valérie Plante Is Breaking Another Gender Barrier In Montreal Politics
Canadian Politicians Are Spreading Lies About Weed
Schedule For All The Christmas Movies On CBC This Holiday Season
You Can Now Get "Fish Cone" Ice Cream In Montreal
This Is Montreal's All-New "Wi-Fi Metro Map"
Downtown Montreal's "Free EDM Festival" 2017 Lineup Announced
Montreal Hosting Massive Scottish Festival This August
Canada Releases First-Ever "Pure Gold" Loonie
More Than Half Of Montreal's STM Metro Stations Will Soon Have Wi-Fi
The Real 1960s "Batmobile" Spotted Driving Downtown Montreal
You Can Now Fly From Montreal To Hong Kong From Only $499 Round Trip
You Can Now Get "Pitch Black/Green Matcha" Swirl Ice Cream In Montreal
It's Official, Dwayne 'The Rock' Johnson Drafted To Run For President In 2020
Montreal Opens All-New "Hawaiian Themed" Restaurant On Saint-Catherine Street
Warning: Montreal Will Be Hit With 7 Straight Days Of Rain
Montreal Is Shutting Down For A Massive "Electro Parade" This Summer
Canadian Students Are Going To Pay A Lot More For School This Year
8 "Horror Movies" You Need To See At Montreal's 2017 "Fantasia International Film Festival"
You Can Now Get $50 Bottle Service In Montreal
Joe Jonas Spotted With His Girlfriend In Montreal
You Can Now Fly Montreal To Japan For $499 Round Trip
This "Lagoon City" Near Montreal Will Make You Feel Like You're In Venice
This Video Of People From USA Visiting Montreal For First Time Is Simply Amazing
8 Awesome Montreal West Island "Hotspots" You Might Not Know About
Montreal Now Has Its First-Ever "Tree Labyrinth"
Montreal's All-New 3.8 km "River-Mountain Walkway" Opens Monday
Quebec Was Hit With A Snowfall Yesterday (Video)
Montreal's Olympic Stadium Will Be Transformed Into An "Urban Campsite" This Summer
This Montreal-Area Starbucks Ranked #1 "Most Beautiful In The World"
The "Hottest" Photos From Montreal Comiccon 2017
You Can Now Get "Churro Poutine" In Montreal
It's Official, Montreal's STM AZUR Metro Trains Are Now Running On The Green Line
Sophie Turner Spotted Montreal Posing For Photos With Every Fan She Meets
This Just In, Justin Trudeau Publicly Bashing French Canadians
It's Official, Deadmau5 Is Coming To Montreal
Here's Why It's Been Raining So Much In Montreal This Summer
Montreal Hosting A "Harry Potter" Theme Brunch
This Is Montreal's All-New "Free Wi-Fi Map", And How To Connect To It
It's Official, Montreal Is Getting Free Wi-Fi All Over The City
Kendrick Lamar Is Opening Pop-Up Shop In Montreal
Montreal Chinatown Hosting A Giant Sidewalk Sale This Summer
Montreal's Belmont Park Is Back
Montreal Hosting A Caribbean Food Festival This Summer
You Can Now Get "Watermelon Bowl" Ice Cream In Montreal
It's Official, Jay-Z Is Coming To Montreal
All The Photos From Montreal’s Comiccon 2017 You Need To See To Believe
Olivia Munn Spotted In Montreal
You Can Now Get 10L Sangria Pitchers In Montreal
Montreal Opening All-New "Moroccan Themed" Restaurant And It's Lit AF
Montreal Is Hosting A Massive "Riverdale" Themed Party
Montreal Will Be Hosting A "Giant Balloon Parade" This Summer
Everyone Is Talking About The "Mysterious Rainbow People" Taking Over Montreal, Here's What's Actually Happening
Montreal Is Having A Huge 70% Off Designer Shoe Sale This Weekend
Montreal Is Hosting A Massive "Mexican Food" Festival
13 "Value Sets" You Can Get At Sephora For Under $13 Right Now
Here's All The New Rules & Changes Being Implemented At Osheaga This Year
Montreal Is Hosting A Massive "Vegan BBQ" Festival
It's Official, You Can Now Do Yoga On Top Of The Montreal Observatory
Katie Holmes Spotted In Montreal
This Montreal Cafe Now Serves "Hot Pink" Lattes
These 5 Montrealers Are Biking 10,000 km Across Europe To Raise Money For Their Best Friend With MS
Montreal Is Hosting A Massive "Harry Potter" Live Concert This July
It's Official, You Can Now Rent BIXI Bikes With Your OPUS Card In Montreal
Montreal's Saint-Catherine Street Hosting Huge "Street Food" Festival Next Week
You Can Now Fly Montreal To Germany For $150
Montreal Is Having A Huge "Mini Cactus" Sale This Weekend
Quebec Ranked "Most Easily Offended" Province In Canada, And We Think That's Offensive
Netflix Canada "Originals" July 2017 Movie & TV Shows Announced

code.py
import numpy
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import LSTM
from keras.callbacks import ModelCheckpoint
from keras.utils import np_utils

# load ascii text and covert to lowercase
filename = "lifestyle.txt"
print("File is: ", filename)
raw_text = open(filename, encoding='UTF8').read()
raw_text = raw_text.lower()

#Map char to ints in a dict
chars = sorted(list(set(raw_text)))
char_to_int = dict((c, i) for i, c in enumerate(chars))

#Grab data info
n_chars = len(raw_text)
n_vocab = len(chars)
print("Total Characters: ", n_chars)
print("Total Vocab: ", n_vocab)

#Sliding window for the input 
seq_length = 100
dataX = []
dataY = []
for i in range(0, n_chars - seq_length, 1):
	seq_in = raw_text[i:i + seq_length]
	seq_out = raw_text[i + seq_length]
	dataX.append([char_to_int[char] for char in seq_in])
	dataY.append(char_to_int[seq_out])
n_patterns = len(dataX)
print("Total Patterns: ", n_patterns)

#Reshape and normalize
X = numpy.reshape(dataX, (n_patterns, seq_length, 1))
X = X / float(n_vocab)
# one hot encode the output variable
y = np_utils.to_categorical(dataY)

#RNN model with LSTM. 2 layers, make sure return_sequences = TRUE
model = Sequential()
model.add(LSTM(256, input_shape=(X.shape[1], X.shape[2]), return_sequences=True))
model.add(Dropout(0.2))
model.add(LSTM(256))
model.add(Dropout(0.2))
model.add(Dense(y.shape[1], activation='softmax'))

#For resuming any training
#filename = "weights\\Newsweights-improvement-16-1.5702-bigger.hdf5"
#model.load_weights(filename)

model.compile(loss='categorical_crossentropy', optimizer='adam')
#Checkpoint
filepath="weights\\lifestyle-improvement-{epoch:02d}-{loss:.4f}-bigger.hdf5"
checkpoint = ModelCheckpoint(filepath, monitor='loss', verbose=1, save_best_only=True, mode='min')
callbacks_list = [checkpoint]

#Fit the model
model.fit(X, y, epochs=1000, batch_size=64, callbacks=callbacks_list)

learner.py
import numpy
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import LSTM
from keras.callbacks import ModelCheckpoint
from keras.utils import np_utils

# load ascii text and covert to lowercase
filename = ""
print("File is: ", filename)
raw_text = open(filename, encoding='UTF8').read()
raw_text = raw_text.lower()

#Map char to ints in a dict
chars = sorted(list(set(raw_text)))
char_to_int = dict((c, i) for i, c in enumerate(chars))

#Grab data info
n_chars = len(raw_text)
n_vocab = len(chars)
print("Total Characters: ", n_chars)
print("Total Vocab: ", n_vocab)

#Sliding window for the input 
seq_length = 100
dataX = []
dataY = []
for i in range(0, n_chars - seq_length, 1):
	seq_in = raw_text[i:i + seq_length]
	seq_out = raw_text[i + seq_length]
	dataX.append([char_to_int[char] for char in seq_in])
	dataY.append(char_to_int[seq_out])
n_patterns = len(dataX)
print("Total Patterns: ", n_patterns)

#Reshape and normalize
X = numpy.reshape(dataX, (n_patterns, seq_length, 1))
X = X / float(n_vocab)
# one hot encode the output variable
y = np_utils.to_categorical(dataY)

#RNN model with LSTM. 2 layers, make sure return_sequences = TRUE
model = Sequential()
model.add(LSTM(256, input_shape=(X.shape[1], X.shape[2]), return_sequences=True))
model.add(Dropout(0.2))
model.add(LSTM(256))
model.add(Dropout(0.2))
model.add(Dense(y.shape[1], activation='softmax'))

#For resuming any training
#filename = "weight-you-want-to-resume.hdf5"
#model.load_weights(filename)

model.compile(loss='categorical_crossentropy', optimizer='adam')
#Checkpoint
filepath="path-to-weights-with-improvement-values-{epoch:02d}-{loss:.4f}-bigger.hdf5"
checkpoint = ModelCheckpoint(filepath, monitor='loss', verbose=1, save_best_only=True, mode='min')
callbacks_list = [checkpoint]

#Fit the model
model.fit(X, y, epochs=1000, batch_size=64, callbacks=callbacks_list)


generator.py
import sys
import numpy
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import LSTM
from keras.callbacks import ModelCheckpoint
from keras.utils import np_utils
# load ascii text and covert to lowercase
filename = ""
raw_text = open(filename, encoding='UTF8').read()
raw_text = raw_text.lower()
# create mapping of unique chars to integers, and a reverse mapping
chars = sorted(list(set(raw_text)))
char_to_int = dict((c, i) for i, c in enumerate(chars))
int_to_char = dict((i, c) for i, c in enumerate(chars))
# summarize the loaded data
n_chars = len(raw_text)
n_vocab = len(chars)
print("Total Characters: ", n_chars)
print("Total Vocab: ", n_vocab)
# prepare the dataset of input to output pairs encoded as integers
seq_length = 100
dataX = []
dataY = []
for i in range(0, n_chars - seq_length, 1):
	seq_in = raw_text[i:i + seq_length]
	seq_out = raw_text[i + seq_length]
	dataX.append([char_to_int[char] for char in seq_in])
	dataY.append(char_to_int[seq_out])
n_patterns = len(dataX)
print("Total Patterns: ", n_patterns)
# reshape X to be [samples, time steps, features]
X = numpy.reshape(dataX, (n_patterns, seq_length, 1))
# normalize
X = X / float(n_vocab)
# one hot encode the output variable
y = np_utils.to_categorical(dataY)
# define the LSTM model
model = Sequential()
model.add(LSTM(256, input_shape=(X.shape[1], X.shape[2]), return_sequences=True))
model.add(Dropout(0.2))
model.add(LSTM(256))
model.add(Dropout(0.2))
model.add(Dense(y.shape[1], activation='softmax'))
# load the network weights
filename = "weight-to-load.hdf5"
model.load_weights(filename)
model.compile(loss='categorical_crossentropy', optimizer='adam')
# pick a random seed
start = numpy.random.randint(0, len(dataX)-1)
pattern = dataX[start]
print("Seed:")
print ("\"", ''.join([int_to_char[value] for value in pattern]), "\"" )
# generate characters
for i in range(300):
	x = numpy.reshape(pattern, (1, len(pattern), 1))
	x = x / float(n_vocab)
	prediction = model.predict(x, verbose=0)
	index = numpy.argmax(prediction)
	result = int_to_char[index]
	seq_in = [int_to_char[value] for value in pattern]
	sys.stdout.write(result)
	pattern.append(index)
	pattern = pattern[1:len(pattern)]
print("\nDone.")
